mysql/resource_ti_config_variable.go

```diff
@@ -0,0 +1,171 @@
+package mysql
+
+import (
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"log"
+	"regexp"
+	"strings"
+
+	"github.com/creasty/defaults"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/tidwall/gjson"
+)
+
+func resourceTiConfigVariable() *schema.Resource {
+	return &schema.Resource{
+		Create: CreateOrUpdateConfigVariable,
+		Read:   ReadConfigVariable,
+		Update: CreateOrUpdateConfigVariable,
+		Delete: DeleteConfigVariable,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"value": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"type": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringInSlice([]string{"pd", "tikv"}, true),
+			},
+			"instance": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+		},
+	}
+}
+
+func CreateOrUpdateConfigVariable(d *schema.ResourceData, meta interface{}) error {
+	db := meta.(*MySQLConfiguration).Db
+
+	varName := d.Get("name").(string)
+	varValue := d.Get("value").(string)
+	varInstanceType := d.Get("type").(string)
+	varInstance := d.Get("instance").(string)
+
+	var warnLevel, warnMessage string
+	var warnCode int = 0
+
+	configQuery := fmt.Sprintf("SET CONFIG %s %s=", varInstanceType, quoteIdentifier(varName))
+
+	if varInstance != "" {
+		configQuery = fmt.Sprintf("SET CONFIG \"%s\" %s=", varInstance, quoteIdentifier(varName))
+	}
+
+	if isNumeric(varValue) {
+		configQuery = fmt.Sprintf("%s %s", configQuery, varValue)
+	} else {
+		configQuery = fmt.Sprintf("%s'%s'", configQuery, varValue)
```

Comment:
What if `varValue` contains apostrophes? Even spitting out a warning is better than nothing.

Reply:
Here is an example If subsystem doesn't accept apostrophes

`mysql> set config pd `log.level`="`warn`";
Query OK, 0 rows affected, 1 warning (0.01 sec)

mysql> show warnings;
+---------+------+---------------------------------------------------------------------------------------+
| Level   | Code | Message                                                                               |
+---------+------+---------------------------------------------------------------------------------------+
| Warning | 1105 | bad request to http://127.0.0.1:2379/pd/api/v1/config: "log level `warn` is illegal"
 |
+---------+------+---------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
`

It will be later rejected by warning check.

I'll add extra logging about this

Reply:
Ok, my question was maybe laid incorrectly.

Can it happen, that some value will contain apostrophe = `'` character? In that case, your SQL will be invalid. BTW your new check for backticks `` ` `` seems to be broken (like it will warn almost every single time).

For example, is it possible or likely, that a human will want to set some value to `something'somethingelse`? If that is possible, then we should certainly add some escaping.

Reply:
Thank you for pointing out this. I added validator for this.